### PR TITLE
[SYCLomatic] Cleanup dpct::segmented_reduce only supports DPC++ native binary operator warning

### PR DESCRIPTION
**Pull request by yihwang-nv**
- Add new value `user-defined-reductions` for `--use-experimental-features` option.
- Use SYCL experimental feature `SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS` when `user-defined-reductions`  enabled.

Signed-off-by: Wang, Yihan <yihan.wang@intel.com>

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]

File: clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h [hunk 1]
```diff
@@ -591,7 +592,7 @@ namespace device {
 namespace detail {
 
 template <typename... _Args> constexpr auto __joint_reduce(_Args... __args) {
-  return sycl::joint_reduce(__args...);
+  return sycl::ext::oneapi::experimental::joint_reduce(__args...);
```
Comment (YuriPlyakhin) [inline clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h]: So, after this change, header files will not compile with the standard SYCL compiler, right?. Using experimental features by default (not under the flag) doesn't look good.....

Comment (yihwang-nv) [reply clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h]: I'll try to use this experimental features by ` --use-experimental-features=`, and add an new value for this option.
```
 --use-experimental-features=<value>     - Comma separated list of experimental features to be used in migrated code.
                                            By default, experimental features will not be used in migrated code.
                                            The values are:
    =free-function-queries                -   Experimental extension that allows getting `id`, `item`, `nd_item`, `group`, and
                                              `sub_group` instances globally.
    =local-memory-kernel-scope-allocation -   Experimental extension that allows allocation of local memory objects at the kernel
                                              functor scope
    =logical-group                        -   Experimental helper function used to logically group work-items.
    =nd_range_barrier                     -   Experimental helper function used to help cross group synchronization during migration.
```

Comment (yihwang-nv) [reply clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h]: > I'll try to use this experimental features by ` --use-experimental-features=`, and add an new value for this option.
> 
> ```
>  --use-experimental-features=<value>     - Comma separated list of experimental features to be used in migrated code.
>                                             By default, experimental features will not be used in migrated code.
>                                             The values are:
>     =free-function-queries                -   Experimental extension that allows getting `id`, `item`, `nd_item`, `group`, and
>                                               `sub_group` instances globally.
>     =local-memory-kernel-scope-allocation -   Experimental extension that allows allocation of local memory objects at the kernel
>                                               functor scope
>     =logical-group                        -   Experimental helper function used to logically group work-items.
>     =nd_range_barrier                     -   Experimental helper function used to help cross group synchronization during migration.
> ```
Maybe
```
--use-experimental-features=user-defined-reductions
```

File: clang/test/dpct/cub/devicelevel/devicesegmentedreduce_user_define_reductions.cu [hunk 2]
```diff
@@ -0,0 +1,591 @@
+// UNSUPPORTED: cuda-8.0, cuda-9.0, cuda-9.1, cuda-9.2, cuda-10.0, cuda-10.1, cuda-10.2
+// UNSUPPORTED: v8.0, v9.0, v9.1, v9.2, v10.0, v10.1, v10.2
+// RUN: dpct --format-range=none -in-root %S -out-root %T/devicelevel/devicesegmentedreduce_user_define_reductions %S/devicesegmentedreduce_user_define_reductions.cu --use-experimental-features=user-defined-reductions --cuda-include-path="%cuda-path/include" -- -std=c++14 -x cuda --cuda-host-only
+// RUN: FileCheck --input-file %T/devicelevel/devicesegmentedreduce_user_define_reductions/devicesegmentedreduce_user_define_reductions.dp.cpp %s --check-prefixes=CHECK,CHECK-DPCT1092
+// RUN: FileCheck --input-file %T/devicelevel/devicesegmentedreduce_user_define_reductions/devicesegmentedreduce_user_define_reductions.dp.cpp %s --check-prefixes=CHECK,CHECK-DPCT1026
+
```
Comment (intwanghao) [inline clang/test/dpct/cub/devicelevel/devicesegmentedreduce_user_define_reductions.cu:6]: Where the macro SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS inserted?

File: clang/runtime/dpct-rt/include/dpl_extras/dpcpp_extensions.h.inc [hunk 3]
```diff
@@ -654,7 +658,11 @@ namespace detail {
 // DPCT_DEPENDENCY_EMPTY
 // DPCT_CODE
 template <typename... _Args> constexpr auto __joint_reduce(_Args... __args) {
+#ifdef SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS 
+  return sycl::ext::oneapi::experimental::joint_reduce(__args...);
```
Comment (intwanghao) [inline clang/runtime/dpct-rt/include/dpl_extras/dpcpp_extensions.h.inc]: Does sycl::ext::oneapi::experimental::joint_reduce compatible with sycl::joint_reduce?
If both user defined op and cub op are used, then all func call will go through first code path.

File: clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h [hunk 4]
```diff
@@ -617,6 +625,10 @@ void segmented_reduce(sycl::queue queue, T *inputs, T *outputs,
   sycl::range<1> local_size(GROUP_SIZE);
 
   queue.submit([&](sycl::handler &cgh) {
+#ifdef SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS
```
Comment (intwanghao) [inline clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h]: Could we not use extra mem for sycl native op?

File: clang/runtime/dpct-rt/include/dpl_extras/dpcpp_extensions.h.inc (hunk 3 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpl_extras/dpcpp_extensions.h.inc]: Yes, `sycl::ext::oneapi::experimental::joint_reduce` compatible with `sycl::joint_reduce`, as far as I know,   if the binary operator is sycl native operator, `sycl::ext::oneapi::experimental::joint_reduce` will call `sycl::joint_reduce` .

File: clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h (hunk 4 above)
Comment (yihwang-nv) [reply clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h]: Okay, I'll optimize this, this extra memory is useless for sycl native operator.

File: clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h [hunk 5]
```diff
@@ -591,9 +595,31 @@ namespace device {
...
 
 template <typename... _Args> constexpr auto __joint_reduce(_Args... __args) {
+#ifdef SYCL_EXT_ONEAPI_USER_DEFINED_REDUCTIONS 
+  return sycl::ext::oneapi::experimental::joint_reduce(__args...);
+#else
   return sycl::joint_reduce(__args...);
+#endif
 }
 
+template <typename _Tp, typename... _Ts>
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h]: 1. Please don't make experimental API by default on. 2. put expermental API based code into experimental:: namespece.   Suggest introduce a new segmented_reduce function into experimental:: namespace and put all impment there.

**Review by zhiweij1 — APPROVED** [review]

File: clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h (hunk 5 above)
Comment (yihwang-nv) [reply clang/test/dpct/helper_files_ref/include/dpl_extras/dpcpp_extensions.h]: done

**Review by ziranzha — APPROVED** [review]

**Review by intwanghao — APPROVED** [review]
LGTM

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 6]
```diff
@@ -369,6 +369,9 @@ DPCT_ENUM_OPTION(DPCT_OPT_TYPE(static llvm::cl::bits<ExperimentalFeatures>), Exp
         false),
              DPCT_OPT_ENUM("nd_range_barrier", int(ExperimentalFeatures::Exp_NdRangeBarrier),
         "Experimental helper function used to help cross group synchronization during migration.\n",
+        false),
+             DPCT_OPT_ENUM("user-defined-reductions", int(ExperimentalFeatures::Exp_UserDefineReductions),
+        "Experimental helper function used to allows user define reductions during migration.\n",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]:
```suggestion
        "Experimental extension that allows user define reductions.\n",
```

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**